src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -941,18 +935,39 @@ fn anchor_failure(
     path_str: &str,
     dox: &str,
     link_range: Option<Range<usize>>,
-    msg: &str,
+    failure: AnchorFailure,
 ) {
-    build_diagnostic(
-        cx,
-        item,
-        path_str,
-        dox,
-        link_range,
-        "has an issue with the link anchor.",
-        msg,
-        None,
-    );
+    let msg = match failure {
+        AnchorFailure::MultipleAnchors => format!("`{}` contains multiple anchors", path_str),
+        AnchorFailure::Primitive
+        | AnchorFailure::Variant
+        | AnchorFailure::AssocConstant
+        | AnchorFailure::AssocType
+        | AnchorFailure::Field
+        | AnchorFailure::Method => {
+            let kind = match failure {
+                AnchorFailure::Primitive => "primitive type",
+                AnchorFailure::Variant => "enum variant",
+                AnchorFailure::AssocConstant => "associated constant",
+                AnchorFailure::AssocType => "associated type",
+                AnchorFailure::Field => "field",
```

Comment:
This no longer has the word `struct` anywhere, which I think was helpful. It's not clear in this context what 'field' means.
```suggestion
                AnchorFailure::Field => "struct field",
```
See https://github.com/rust-lang/rust/pull/74528/files#diff-fc441bb62b41117616a86bd40c6f2e5eR1 for an example, that error message is no longer very clear.

Reply:
Fixed.